packages/core/src/utils/userop.ts

```diff
@@ -72,3 +144,183 @@ export function applyUserOpOverrideOrFeeOption(
+  const encoded = encodeAbiParameters(
+    [{ type: "bytes32" }, { type: "address" }, { type: "uint256" }],
+    [
+      keccak256(packUserOperation<TEntryPointVersion>(request)),
+      entryPointAddress,
+      BigInt(chainId),
+    ]
+  ) as `0x${string}`;
+
+  return keccak256(encoded);
+}
+
+/**
+ * Pack the user operation data into bytes for hashing for entry point version 0.6 onwards
+ * Reference:
+ * v0.6.0: https://github.com/eth-infinitism/account-abstraction/blob/releases/v0.6/test/UserOp.ts#L16-L61
+ * v0.7.0: https://github.com/eth-infinitism/account-abstraction/blob/releases/v0.7/test/UserOp.ts#L28-L67
+ *
+ * @param request - the UserOperation to get the hash for
+ * @param entryPointVersion - a {@link EntryPointVersion} entry point contract version
+ * @returns the hash of the UserOperation
+ */
+export function packUserOperation<
+  TEntryPointVersion extends EntryPointVersion = EntryPointVersion
+>(request: UserOperationRequest<TEntryPointVersion>): Hex {
+  switch ("paymaster" in request ? "0.7.0" : "0.6.0") {
+    case "0.7.0":
+      const initCode = concat([
+        (request as UserOperationRequest_v7).factory,
+        (request as UserOperationRequest_v7).factoryData,
+      ]);
+      const accountGasLimits = packAccountGasLimits(
+        (({ verificationGasLimit, callGasLimit }) => ({
+          verificationGasLimit,
+          callGasLimit,
+        }))(request as UserOperationRequest<"0.7.0">)
+      );
+
+      const gasFees = packAccountGasLimits(
+        (({ maxPriorityFeePerGas, maxFeePerGas }) => ({
+          maxPriorityFeePerGas,
+          maxFeePerGas,
+        }))(request as UserOperationRequest<"0.7.0">)
+      );
+
+      const paymasterAndData = isAddress(
+        (request as UserOperationRequest<"0.7.0">).paymaster
+      )
+        ? packPaymasterData(
+            (({
+              paymaster,
+              paymasterVerificationGasLimit,
+              paymasterPostOpGasLimit,
+              paymasterData,
+            }) => ({
+              paymaster,
+              paymasterVerificationGasLimit,
+              paymasterPostOpGasLimit,
+              paymasterData,
+            }))(request as UserOperationRequest<"0.7.0">)
+          )
+        : "0x";
+
+      return encodeAbiParameters(
+        [
+          { type: "address" },
+          { type: "uint256" },
+          { type: "bytes32" },
+          { type: "bytes32" },
+          { type: "bytes32" },
+          { type: "uint256" },
+          { type: "bytes32" },
+          { type: "bytes32" },
+        ],
+        [
+          request.sender as Address,
+          hexToBigInt(request.nonce),
+          keccak256(initCode),
+          keccak256(request.callData),
+          accountGasLimits,
+          hexToBigInt(request.preVerificationGas),
+          gasFees,
+          paymasterAndData,
+        ]
+      );
+
+    case "0.6.0":
+      const hashedInitCode = keccak256(
+        (request as UserOperationRequest<"0.6.0">).initCode
```

Comment:
going off the above comment. it would omit having to do this `as` since the EP itself parses the UO for validity before parsing it